src/Features/LanguageServer/Protocol/Features/EditAndContinue/EditAndContinueDiagnosticSource.cs

```diff
@@ -2,19 +2,77 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.EditAndContinue;
 
 internal sealed class EditAndContinueDiagnosticSource
 {
-    public static async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(Document designTimeDocument, CancellationToken cancellationToken)
+    private sealed class ClosedDocumentSource(TextDocument document, ImmutableArray<DiagnosticData> diagnostics) : AbstractWorkspaceDocumentDiagnosticSource(document)
+    {
+        public override bool IsLiveSource()
+            => true;
+
+        public override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(IDiagnosticAnalyzerService diagnosticAnalyzerService, RequestContext context, CancellationToken cancellationToken)
+            => Task.FromResult(diagnostics);
+    }
+
+    private sealed class ProjectSource(Project project, ImmutableArray<DiagnosticData> diagnostics) : AbstractProjectDiagnosticSource(project)
+    {
+        public override bool IsLiveSource()
+            => true;
+
+        public override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(IDiagnosticAnalyzerService diagnosticAnalyzerService, RequestContext context, CancellationToken cancellationToken)
+            => Task.FromResult(diagnostics);
+    }
+
+    public static async ValueTask AddWorkspaceDiagnosticSourcesAsync(ArrayBuilder<IDiagnosticSource> sources, Solution solution, Func<Document, bool> isDocumentOpen, CancellationToken cancellationToken)
+    {
+        // avoid creating and synchronizing compile-time solution if Hot Reload/EnC session is not active
+        if (solution.Services.GetRequiredService<IEditAndContinueWorkspaceService>().SessionTracker is not { IsSessionActive: true } sessionStateTracker)
+        {
+            return;
+        }
+
+        var applyDiagnostics = sessionStateTracker.ApplyChangesDiagnostics;
+
+        var dataByDocument = from data in applyDiagnostics
+                             where data.DocumentId != null
+                             group data by data.DocumentId into documentData
+                             select documentData;
+
+        // diagnostics associated with closed documents:
```

Comment:
is the idea that live diagnostics already handles thsi case?

Note: would it actually be cleaner to have EnC add *all* its sources here (including for open documents), and have the normal sources not et the EnCDiagnostics at all?

Reply:
Aren't workspace pull diagnostics pulled less frequently? The open document diagnostics should be reported as the user types.

Reply:
> Is the idea that live diagnostics already handles thsi case?

Yes

Reply:
Sorry, my point is: You should add your own open document diagnostic handler which is pulled on at the normal frequesncy.  But you then own the entire diagnostics flow from EnC to the LSP handler, without having some of it having to go through this path and some of it going though DiagAnalyzerService.

Reply:
By way of analogy.  THis is how the task/todo-list support works now in pull diagnostics.  There's a dedicated DiagSource for todo-items, both for open docs, and for the closed workspace docs.  I'm just thinking the same might make sense for you for EnC document diagnostics :)

Reply:
Note: i would be very ok with us making this extensible so that TaskList knowledge doesn't live in the pull-diags code, and it instead discovers thsi through some mechanism to allow those systems to just add their diag sources through MEF.  tehn we can have the todo/tasklist code move to a dedicated location.

Reply:
I tried that but it does not work with the current design. The pull diagnostic handler currently expects at most one `IDiagnosticSource` per `TextDocument`. I can't add another one without breaking a lot of things.  @dibarbet

Reply:
Unless we add a new diagnostic category: `PullDiagnosticCategories.EnC` similarly to `PullDiagnosticCategories.Task`